### ISSUE TITLE
After cancelling all subscriptions of Observable

### ISSUE THREAD

**Issue by hironytic** [ReactiveX/rxdart#203]
In RxDart, I believe `PublishSubject` is a hot observable, and the code below works fine as I expect.

```dart
final subject = PublishSubject<int>();
final Observable<int> observable = subject;
// final Observable<int> observable = subject.shareReplay();

final subscriptionX = observable.listen((value) => print("X: $value"));
final subscriptionY = observable.listen((value) => print("Y: $value"));

subject.add(1);
subject.add(2);

// Wait to avoid canceling subscriptions before callbacks are called
await Future.delayed(Duration(seconds: 1));
await subscriptionX.cancel();
await subscriptionY.cancel();

final subscriptionZ = observable.listen((value) => print("Z: $value"));

subject.add(3);

await Future.delayed(Duration(seconds: 1));
await subscriptionZ.cancel();

await subject.close();

// output
// 
// X: 1
// Y: 1
// X: 2
// Y: 2
// Z: 3
```

Then I add `shareReplay` operator to the subject, past values are replayed as I expect but the new value (`3`) is not appeared on subscription Z.

```dart
// output
//
// X: 1
// Y: 1
// X: 2
// Y: 2
// Z: 1
// Z: 2
```

I feel this is an inconsist behavior.
Once all subscriptions of an observable are cancelled and starting to listen it again, is the behavior of such a case undefined in RxDart?

**Comment by brianegan** [comment]
Hey there -- this is a fair point. Rather than only describe the problem, could you please describe what you're hoping to achieve? I'm wondering if there's a different way to tackle the problem.

In general, these operators are built to turn a Single-Subscription Stream (kind of like a "Cold" Observable) into a Broadcast Stream (kind of like a "Hot" observable). 

They are not built to convert one kind of Broadcast Stream into another. Since "Cold" Streams in Dart are single-subscription, the `share` operator does not reconnect to the underlying Stream after it's disconnected from the Stream because that will result in a `StateError`. 

To be honest, I actually wonder if that is a fundamental flaw of making Dart streams work for this use-case :/

**Comment by hironytic** [comment]
@brianegan 
Thanks, you are right.

I wanted to find out the way to "share" an observable regardless of whether it is hot or cold.

Suppose I am making an input form and writing its business logic with RxDart.
Each time user inputs, its value is come to `loginName`.
It is automatically validated. An error message may appear in screen, and the submit button changes to enabled or disabled. These are outputs.

```dart
// This is user's input.
Observable<String> loginName = ...

Observable<bool> validationResult = loginName.map((value) {
  // This returns false when validation failed.
  return validateLoginName(value);
});

// This is output. Showing error message when validation failed.
Observable<String> errorMessage = validationResult.map((result) {
  return (result) ? "" : "Validation error";
});

// This is also output. Submit button is only enabled when validation passed.
Observable<bool> isSubmitButtonDisabled = validationResult.map((result) {
  return !result;
});
```

When I posted this issue, I believed in that the `map` operator created a cold observable like other Rx. (That was my misunderstanding, right?)

Because `validationResult` is shared by two observables, I think I need to convert it to hot one. Additionally, if binding the output streams to UI are delayed it should use the last validation result, I think. So I added `shareReplay(maxSize: 1)` to `validationResult`.
And actually I passed the `PublishSubject` to `loginName`.

It worked fine until unbinding these outputs to UI.
It's my background.

**Comment by hironytic** [comment]
Hi there, my original problem was to find out the way to "share" the observable because two or more listeners might subscribe it. One I thought `shareReplay` was the answer, but the behavior was changed with it after canceling all subscriptions. So I posted this issue.
But wait. The original problem came from my misunderstanding: The `map` operator returns a "Cold" observable (single-subscription stream).

In other Rx implementations, most operators, like `map`, always create a "Cold" observable although it can be subscribed by two or more listeners. When two listeners are subscribing it, the mapper passed to `map` is called two times. This situation is "not shared."
In RxDart, an observable with single-subscription stream can be subscribed by only one listener. A broadcast stream are needed to "share" the observable. But, if I understood correctly, most operators don't change the broadcasting status.  The result of the `map` of broadcast stream is a broadcast stream. It can be subscribed by multiple listeners, though the mapper is called two or more times (I don't know whether it is called "Hot" or "Cold.")

Then back to my original problem, @brianegan 's comment makes sense for me.

> In general, these operators are built to turn a Single-Subscription Stream (kind of like a "Cold" Observable) into a Broadcast Stream (kind of like a "Hot" observable).

I don't need to "share" 🙂 
So I would happy to close this issue unless there is anyone wants to have more discussion about this.
